bin/i18n/translation_monitor/crowdin_validator.rb

```diff
@@ -0,0 +1,348 @@
+                'source_string_count' => source_strings.size,
+                'error_count' => errors.size
+              }
+            }
+          )
+          key = create_history_key(config)
+          history[key] = config_with_results
+          write_to_json(history, history_file)
+        end
+      rescue Exception => e
+        puts "Error: #{e.message}"
+        puts e.backtrace
+      end
+    end
+  end
+
+  # Run a single configuration.
+  #
+  # This function has side effects. It updates the +config+ input.
+  # It may write to the local file system and Google Drive.
+  #
+  # @param config [Hash]
+  # @param download_from_crowdin [Boolean] downloading data from Crowdin or reading data from the local file system
+  # @return [Array<Array>] array of translations, source_strings, and translation errors
+  #
+  def run_config(config, download_from_crowdin)
+    missing_params = find_missing_params(config, REQUIRED_CONFIG_PARAMS)
+    raise "Missing config params #{missing_params}" unless missing_params.empty?
+    infer_config_params(config)
+
+    # Retreive translations
+    if download_from_crowdin
+      translations = @crowdin_client.download_translations(
+        config['project_name'],
+        config['crowdin_language_id'],
+        config['user_name'],
+        config['start_date'],
+        config['end_date']
+      )
+      if config['write_to_file']
+        write_to_json(translations, config['translations_json'])
+        puts "Wrote #{translations.size} translations to #{config['translations_json']}"
+      end
+    else
+      translations = read_from_json(config['translations_json'])
+    end
+
+    # Retreive source strings
+    if download_from_crowdin
+      source_strings = @crowdin_client.download_source_strings(
+        config['project_name'],
+        config['crowdin_language_id'],
+        config['user_name'],
+        config['start_date'],
+        config['end_date']
+      )
+      if config['write_to_file']
+        write_to_json(source_strings, config['source_strings_json'])
+        puts "Wrote #{source_strings.size} source strings to #{config['source_strings_json']}"
+      end
+    else
+      source_strings = read_from_json(config['source_strings_json'])
+    end
+
+    # Find all translation errors
+    errors = validate_all_translations(translations, source_strings)
+
+    if errors.empty?
+      puts "No errors, skip writing to files and ghseet."
+    else
+      error_rows_with_headers = convert_to_csv_rows(errors)
+      if config['write_to_file']
+        write_to_json errors, config['errors_json']
+        puts "Wrote #{errors.size} errors to #{config['errors_json']}"
+        write_to_csv error_rows_with_headers, config['errors_csv']
+        puts "Wrote #{errors.size} errors to #{config['errors_csv']}"
+      end
+      if config['write_to_gsheet']
+        append_to_gsheet error_rows_with_headers, config['errors_gsheet'], config['crowdin_language_id'], GOOGLE_DRIVE_CREDENTIAL_FILE
```

Comment:
Instead of a separate gsheet per User & Locale, how hard would it be to have 1 gsheet per User and each Locale is a tab in the gsheet?

Reply:
Each gsheet is for a vendor + a project, and each tab in a gsheet is a locale.